accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+ldnull
+cgt.un
+call       void [System.Console]System.Console::WriteLine(bool)
+ldc.i4.0
+call       void [System.Console]System.Console::WriteLine(bool)
+```
+
+The C# compiler will also consider the `false` case warning CS0184: `The given expression is never of the provided...`.
+
+This C# vs. runtime disagreement with "known values" also appears with array variance:
+```cs
+Console.WriteLine(new int[1] is uint[]); // false
+Console.WriteLine(((object)new int[1]) is uint[]); // true
+```
+
+### With modifiers + unmodified types
+The above examples only compare parameter types without calling conventions or modifiers.
+
+An unmodified type does include managed vs. unmanaged as part of its type identity:
+```cs
+object o = new delegate*<int>[1];
+Console.WriteLine(o is delegate*<int>[]); // true
+Console.WriteLine(o is delegate* unmanaged[Cdecl]<int>[]); // false
+
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
+```
+
+but unmanaged calling conventions are not part of the type identity:
+```cs
+object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
+Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true
+
+object o2 = new delegate* unmanaged[CDecl]<int>[1];
+Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
+```
+
+In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.
+
+### With modified types
+A modified type is a simple wrapper that does not forward the other various `Type` members to the underlying type. It does, however intercept `GetFunctionPointerParameters()` to return modified type parameters. It also implements `GetOptionalCustomModifiers()` and `GetRequiredCustomModifiers()` which throw on unmodified types.
+
+Thus there isn't a `Type` identity for a modified type since it is created lazily and not cached. Comparing using `==` or `Equals()` will return `false`.
+
+```cs
+public class Holder
+{
+    public unsafe delegate* unmanaged[SuppressGCTransition]<int> _field;
+}
+
+Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType()!;
+Console.WriteLine(modifiedType.IsModifiedType); // true
+Console.WriteLine(modifiedType.UnderlyingSystemType); // (function pointer)
+Console.WriteLine(modifiedType.IsFunctionPointer); // false
+Console.WriteLine(modifiedType.UnderlyingSystemType.IsFunctionPointer); // true
+
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions().Length); // 2
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0]); // CallConvCdecl
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[1]); // CallConvMemberFunction
+
+Type modifiedType2 = typeof(Holder).GetField("_field").GetModifiedFieldType()!; // same field...
+Console.WriteLine(modifiedType == modifiedType2); // false (a different type instance; not cached)
```

Comment:
Are we going to do extra work to make this not work?

`==` is implemented by calling `Equals` and unless we override it, the comparison would be done on `UnderlyingSystemType` and return `true`. I think that might be the desired behavior. It is still the same type, just with extra metadata attached to it.

Reply:
(Or maybe you meant `object.ReferenceEquals` instead of `==`?)

Reply:
Based on other feedback as well around the `GetModified*Type()` methods returning `null` or not, I think it makes the most sense to treat the modified types as essentially the underlying type except with different identity (as expected, since one will have custom modifiers).